### PR TITLE
support filter chain for filtering response data

### LINKED ISSUE #1188: [feature request]support filter chain to mutate response data in YurtHub
**What would you like to be added**:
The response data of request from components(like kube-proxy, etc.) can be filtered by multiple filters in Yurthub. At present, only one filter can be selected by the request.

**Why is this needed**:
As mentioned in issue #1183 , we will a new filter in Yurthub to mutate response for kube-proxy list/watch service request, but `discardcloudservice` filter is also used by kube-proxy list/watch service request, so we need to support filter chain in order to make these two filters selected by kube-proxy list/watch service.

**others**
/kind feature

### PR DESCRIPTION
**Pull request by rambohe-ch**
/kind feature

#### What this PR does / why we need it:
improve filter framework in Yurthub in order to support filter chain that multiple filters can be selected for response data.

#### Which issue(s) this PR fixes:

Fixes #1188 

#### Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: pkg/yurthub/filter/servicetopology/filter.go [hunk 1]
```diff
@@ -174,11 +184,246 @@ func (ssf *serviceTopologyFilter) SetStorageWrapper(s cachemanager.StorageWrappe
 	return nil
 }
 
-func (ssf *serviceTopologyFilter) Filter(req *http.Request, rc io.ReadCloser, stopCh <-chan struct{}) (int, io.ReadCloser, error) {
-	if ok := cache.WaitForCacheSync(stopCh, ssf.nodeSynced, ssf.serviceSynced, ssf.nodePoolSynced); !ok {
-		return 0, rc, nil
+func (stf *serviceTopologyFilter) Filter(obj runtime.Object, stopCh <-chan struct{}) runtime.Object {
+	if ok := cache.WaitForCacheSync(stopCh, stf.nodeSynced, stf.serviceSynced, stf.nodePoolSynced); !ok {
+		return obj
```
Comment (Congrool) [inline pkg/yurthub/filter/servicetopology/filter.go:189]: If stopCh is closed, it seems to return the unfiltered object to the client. Is it suitable?

File: pkg/yurthub/proxy/pool/pool.go [hunk 2]
```diff
@@ -229,7 +229,7 @@ func (pp *PoolCoordinatorProxy) modifyResponse(resp *http.Response) error {
 
 		// filter response data
 		if pp.filterMgr != nil {
-			if ok, runner := pp.filterMgr.FindRunner(req); ok {
+			if runner, ok := pp.filterMgr.FindResponseFilter(req); ok {
```
[the remark this answers is not in the document]
Comment (rambohe-ch) [reply pkg/yurthub/proxy/pool/pool.go:232]: fixed

File: pkg/yurthub/proxy/remote/loadbalancer.go [hunk 3]
```diff
@@ -288,7 +288,7 @@ func (lb *loadBalancer) modifyResponse(resp *http.Response) error {
 
 		// filter response data
 		if lb.filterManager != nil {
-			if ok, runner := lb.filterManager.FindRunner(req); ok {
+			if runner, ok := lb.filterManager.FindResponseFilter(req); ok {
```
[the remark this answers is not in the document]
Comment (rambohe-ch) [reply pkg/yurthub/proxy/remote/loadbalancer.go:291]: fixed

File: pkg/yurthub/filter/filter.go [hunk 4]
```diff
@@ -262,3 +264,62 @@ func CreateSerializer(respContentType string, info *apirequest.RequestInfo, sm *
...
+	// do nothing
+	return map[string]sets.String{}
+}
+
+func (chain filterChain) Filter(obj runtime.Object, stopCh <-chan struct{}) runtime.Object {
+	if len(chain) == 0 {
+		return obj
+	} else if len(chain) == 1 {
+		return chain[0].Filter(obj, stopCh)
+	}
```
[the remark this answers is not in the document]
Comment (rambohe-ch) [reply pkg/yurthub/filter/filter.go:298]: fixed

File: pkg/yurthub/filter/servicetopology/filter.go (hunk 1 above)
Comment (rambohe-ch) [reply pkg/yurthub/filter/servicetopology/filter.go:189]: if stopCh is closed, Yurthub will be exit and restart lately. so it's no problem to return a unfiltered object.

**Review by Congrool — COMMENTED** [review]
/lgtm

**Pull request merged**